app/static/ringing_room.js

```diff
@@ -787,6 +811,489 @@ $(document).ready(function() {
 `,
     }); // End help
 
+    Vue.component('wheatley', {
+        data: function () {
+            return {
+                enabled: false,
+                sensitivity: 0.6,
+                use_up_down_in: true,
+                stop_at_rounds: true,
+
+                row_gen: {
+                    type: "method",
+                    title: "Double Norwich Court Bob Major",
+                    url: "Double_Norwich_Court_Bob_Major",
+                },
+
+                is_ringing: false,
+
+                row_gen_panel: "method",
+
+                method_name: "",
+                autocomplete_options: [],
+                selected_option: 0,
+
+                complib_id: "",
+                current_complib_comp: undefined,
+                complib_error: ""
+            };
+        },
+
+        computed: {
+            display_text: function () {
+                if (this.autocomplete_options.length == 0) {
+                    return "NOWT";
+                }
+
+                return this.autocomplete_options.map(x => x.title).join(", ");
+            },
+            touch_link: function () {
+                let row_gen = this.row_gen;
+
+                if (row_gen) {
+                    if (row_gen.type == "method") {
+                        return "https://rsw.me.uk/blueline/methods/view/" + 
+                            (row_gen.url || "Grandsire_Major");
+                    } else if (row_gen.type == "composition") {
+                        return row_gen.url || "";
+                    } else {
+                        return "";
+                    }
+                } else {
+                    return "";
+                }
+            },
+            touch_text: function () {
+                let row_gen = this.row_gen;
+
+                if (row_gen) {
+                    if (row_gen.type == "method") {
+                        return row_gen.title || "No Method Title";
+                    } else if (row_gen.type == "composition") {
+                        return row_gen.title || "No Composition Title";
+                    } else {
+                        return "<unknown row_gen type " + row_gen.type + ">";
+                    }
+                } else {
+                    return "<no row gen>";
+                }
+            },
+            row_gen_panel_disabled: function () {
+                return bell_circle.lock_controls || this.is_ringing;
+            },
+            settings_panel_disabled: function () {
+                return bell_circle.lock_controls;
+            }
+        },
+
+        watch: {
+            method_name: function (next_value) {
+                this.update_method_suggestions(next_value);
+            },
+            complib_id: function (next_value) {
+                this.update_comp_suggestions(next_value);
```

Comment:
Can we debounce these? Currently it sends a web request for every keystroke which is an unnecessary load on those hosting these apis.

Reply:
For Bob Wallis' blueline, he's agreed that this is totally OK - his website does this anyway so it's just like a few extra users on his site typing keystrokes.

For complib, I'll ask Graham John, but I think I'll probably end up debouncing this anyway.  I'd like to add functionality for allowing people to paste URLs that may include access keys to private compositions, but at the moment I just want to get Wheatley out there and basically functional before I need to start revising for exams.